packages/xforms-engine/src/parse/xpath/dependency-analysis.ts

```diff
@@ -0,0 +1,123 @@
+import { expressionParser } from '@getodk/xpath/expressionParser.js';
+import {
+	resolvePath,
+	resolvePredicateReference,
+	serializeNodesetReference,
+} from './path-resolution.ts';
+import { findPredicateReferences } from './predicate-analysis.ts';
+import { findLocationPathSubExpressionNodes, getPathExpressionNode } from './semantic-analysis.ts';
+
+interface PathResolutionOptions {
+	/**
+	 * Excludes direct references to the nodeset specified as a context. This flag
+	 * has earlier precedent as `ignoreContextReference` in similar internal APIs.
+	 * The intent, in part, is to allow for certain in-spec exceptions to cycle
+	 * analysis (which we previously did explicitly, and foresee reintroducing in
+	 * some form relatively soon). As such, the default for this flag should be
+	 * overridden when a computation is (A) not among the core `<bind>`
+	 * computations or (B) is a computation which is explicitly expected to have
+	 * self-references (such as `constraint`).
+	 *
+	 * @default false
+	 *
+	 * @todo It is highly likely we will eliminate this in the future. We
+	 * considered removing its precursor (and existing code now referencing this)
+	 * in {@link https://github.com/getodk/web-forms/pull/67}, but ran into issues
+	 * with some of the reactive logic introduced there. It's likely that
+	 * subsequent improvements have addressed those issues, but it's also likely
+	 * that we will want similar special casing when we reintroduce explicit graph
+	 * cycle analysis.
+	 */
+	readonly ignoreReferenceToContextPath?: boolean;
+
+	/**
+	 * Treats the literal sub-expression `null` as if it were a keyword. This is a
+	 * deviation from XPath (where `null` would be a RelativeLocationPath
+	 * referencing a child node by that name). However, in real-world usage, some
+	 * forms evidently use `null` as a keyword representing a null and/or blank
```

Comment:
I'm glad you asked about this! It gave me an opportunity to really think about the underlying reason it existed in the first place, and whether it makes sense to carry over into this change. In short: I don't think this is necessary anymore! I'll double check a couple things locally to be absolutely sure, but I expect to be able to remove this.

Since you asked for an example form, the cases I could find were in whova. See [here](https://github.com/getodk/web-forms/blob/cfde5ea53085cdae22ffcfa274cb8a48dfc3d699/packages/scenario/resources/smoketests/whova_form.xml#L14627C1-L14628C1) and [here](https://github.com/getodk/web-forms/blob/cfde5ea53085cdae22ffcfa274cb8a48dfc3d699/packages/scenario/resources/smoketests/whova_form.xml#L14634C1-L14635C1) (last argument to the inner `if()` in both).

- - -

I do have to ask whether a `'null'` string value is expected to be a special case. This would be very surprising to me!

Taking the whova usage as an example, the conditions that reference `null` as a NameTest will produce a blank value when they hit that condition. If those were a `'null'` string Literal, in strict XPath semantics, they'll produce the that literal string value.

A quick search of the `resources` ported from JavaRosa doesn't reveal any usage of `'null'` that I could find. A quick search in JavaRosa itself has a few hits for the string literal, but none that look like treating it as a special XPath value.

Reply:
> See [here](https://github.com/getodk/web-forms/blob/cfde5ea53085cdae22ffcfa274cb8a48dfc3d699/packages/scenario/resources/smoketests/whova_form.xml#L14627C1-L14628C1) and [here](https://github.com/getodk/web-forms/blob/cfde5ea53085cdae22ffcfa274cb8a48dfc3d699/packages/scenario/resources/smoketests/whova_form.xml#L14634C1-L14635C1) (last argument to the inner if() in both).

🤯 Looks like the intent is indeed to have a blank value and because it had the desired outcome for the reasons you described it stayed in.

> I do have to ask whether a 'null' string value is expected to be a special case.

No, I don't think so!

> I expect to be able to remove this.

🎉🎉🎉